File: swift/plugins/Sigv4.swift
// AWS Signature Version 4, hand-rolled.
//
// The AWS providers need exactly one thing from the AWS SDK - request
// signing - and taking the SDK for it would break the no-dependency rule
// that keeps the ports honest. SigV4 is a stable, published algorithm
// built from HMAC-SHA256, which Crypto.swift supplies.
//
// `sigv4` is pure: the caller passes the timestamp, so the same input
// yields the same signature everywhere. That is what lets the shared spec
// carry known-answer cases that all ports must reproduce bit-for-bit, and
// lets the integration mock recompute the signature server-side.
//
// It travels with the aws plugin, not with the core: signing a request is
// what makes a kind a plugin (docs/design/plugin-providers.md).
//
// A port of typescript/src/Sigv4.ts, which is canonical.

import Foundation

import Sekreto

/// One request to sign - the same declarative shape the shared spec uses.
/// `datetime` is `YYYYMMDDTHHMMSSZ`, and it is the caller's, so that
/// signing is a pure function of its input.
public struct Signing {

  public var method: String
  public var url: String
  public var service: String
  public var region: String
  public var keyid: String
  public var secret: String
  public var datetime: String
  public var headers: Ordered<String>
  public var body: String
  public var session: String?

  public init(
    method: String,
    url: String,
    service: String,
    region: String,
    keyid: String,
    secret: String,
    datetime: String,
    headers: Ordered<String> = Ordered<String>(),
    body: String = "",
    session: String? = nil
  ) {
    self.method = method
    self.url = url
    self.service = service
    self.region = region
    self.keyid = keyid
    self.secret = secret
    self.datetime = datetime
    self.headers = headers
    self.body = body
    self.session = session
  }
}

/// SHA-256 of some text, as lowercase hex.
public func sha256hex(_ text: String) -> String {
  return hex(sha256(Array(text.utf8)))
}

/// HMAC-SHA256 of some text under a key.
public func hmac(_ key: [UInt8], _ text: String) -> [UInt8] {
  return hmacsha256(key, Array(text.utf8))
}


/// Percent-decode, and nothing else: `+` stays `+`, as on the wire, and a
/// malformed escape is kept literal.
public func uridecode(_ text: String) -> String {
  let chars = Array(text.utf8)
  var out: [UInt8] = []
  var index = 0

  while index < chars.count {
    var taken = false

    if 0x25 == chars[index], index + 2 < chars.count {  // '%'
      let digits = String(decoding: chars[(index + 1)...(index + 2)], as: UTF8.self)
      if let code = UInt8(digits, radix: 16) {
        out.append(code)
        index += 3
        taken = true
      }
    }

    if !taken {
      out.append(chars[index])
      index += 1
    }
  }

  return String(decoding: out, as: UTF8.self)
}

/// The canonical query string: each pair RFC 3986-escaped, sorted by
/// escaped key then escaped value.
public func canonicalquery(_ query: String) -> String {
  if query.isEmpty { return "" }

  var pairs: [(String, String)] = []

  for pair in query.components(separatedBy: "&") {
    let split = pair.firstIndex(of: "=")
    let key = nil == split ? pair : String(pair[pair.startIndex..<split!])
    let value = nil == split ? "" : String(pair[pair.index(after: split!)...])
    pairs.append((uriescape(uridecode(key)), uriescape(uridecode(value))))
  }

  pairs.sort { left, right in
    if left.0 != right.0 { return left.0 < right.0 }
    return left.1 < right.1
  }

  return pairs.map { "\($0.0)=\($0.1)" }.joined(separator: "&")
}

/// A URL, split by hand.
///
/// Not URLComponents: what SigV4 signs must match what the address check
/// saw and what the client will dial, and a platform URL type is free to
/// normalise any of the three differently. `host` follows the WHATWG rule
/// the AWS SDKs use - lowercased, userinfo stripped, the port present only
/// when it is not the scheme's default.
struct Urlparts {
  var scheme: String
  var host: String
  var path: String
  var query: String
}

func urlparts(_ url: String) -> Urlparts {
  var scheme = ""
  var rest = url

  if let mark = url.range(of: "://") {
    scheme = String(url[url.startIndex..<mark.lowerBound]).lowercased()
    rest = String(url[mark.upperBound...])
  }

  var authority = rest
  var tail = ""

  if let stop = rest.firstIndex(where: { "/" == $0 || "?" == $0 || "#" == $0 }) {
    authority = String(rest[rest.startIndex..<stop])
    tail = String(rest[stop...])
  }

  // Userinfo is not part of the host, and never part of a signature.
  if let at = authority.lastIndex(of: "@") {
    authority = String(authority[authority.index(after: at)...])
  }

  var hostname = authority
  var port = ""

  if authority.hasPrefix("["), let close = authority.firstIndex(of: "]") {
    hostname = String(authority[authority.startIndex...close])
    let after = String(authority[authority.index(after: close)...])
    if after.hasPrefix(":") { port = String(after.dropFirst()) }
  } else if let colon = authority.lastIndex(of: ":") {
    hostname = String(authority[authority.startIndex..<colon])
    port = String(authority[authority.index(after: colon)...])
  }

  hostname = asciilower(hostname)

  // A default port is implicit: `Host: example.com:443` is not what an AWS
  // SDK signs, and a signature over the wrong host is refused.
  if ("https" == scheme && "443" == port) || ("http" == scheme && "80" == port) {
    port = ""
  }

  var path = "/"
  var query = ""

  if !tail.isEmpty {
    var pathpart = tail

    if let hash = tail.firstIndex(of: "#") {
      pathpart = String(tail[tail.startIndex..<hash])
    }

    if let mark = pathpart.firstIndex(of: "?") {
      query = String(pathpart[pathpart.index(after: mark)...])
      pathpart = String(pathpart[pathpart.startIndex..<mark])
    }

    if !pathpart.isEmpty { path = pathpart }
  }

  return Urlparts(
    scheme: scheme,
    host: port.isEmpty ? hostname : hostname + ":" + port,
    path: path,
    query: query
  )
}

/// Trim, and collapse every internal run of spaces and tabs to one space.
///
/// AWS folds sequential whitespace before signing, so a header value like
/// `a  b\tc` must sign as `a b c` or the service refuses the request.
func foldspace(_ text: String) -> String {
  var out = ""
  var pending = false
  var started = false

  for ch in text {
    if " " == ch || "\t" == ch || "\n" == ch || "\r" == ch {
      if started { pending = true }
      continue
    }
    if pending {
      out.append(" ")
      pending = false
    }
    out.append(ch)
    started = true
  }

  return out
}

/// Sign one request. Returns the headers to attach: authorization,
/// x-amz-date, and x-amz-security-token when a session token was given, in
/// that order - the spec compares the result as a JSON object, and callers
/// print it field by field.
public func sigv4(_ input: Signing) -> Ordered<String> {
  let parts = urlparts(input.url)

  let date = String(input.datetime.prefix(8))
  let session = (input.session?.isEmpty ?? true) ? nil : input.session

  // Every header that will be signed: the caller's extras, plus host and
  // x-amz-date (and the session token when present). The built-in three go
  // in AFTER the caller's, so they win over anything passed in.
  var headers = Ordered<String>()

  for (key, value) in input.headers.pairs {
    headers[asciilower(key)] = foldspace(value)
  }

  headers["host"] = parts.host
  headers["x-amz-date"] = input.datetime
  if let token = session { headers["x-amz-security-token"] = token }

  let sorted = headers.pairs.sorted { $0.0 < $1.0 }

  let canonicalheaders = sorted.map { "\($0.0):\($0.1)\n" }.joined()
  let signedheaders = sorted.map { $0.0 }.joined(separator: ";")

  let canonicalrequest = [
    asciiupper(input.method),
    parts.path,
    canonicalquery(parts.query),
    canonicalheaders,
    signedheaders,
    sha256hex(input.body),
  ].joined(separator: "\n")

  let scope = "\(date)/\(input.region)/\(input.service)/aws4_request"

  let stringtosign = [
    "AWS4-HMAC-SHA256",
    input.datetime,
    scope,
    sha256hex(canonicalrequest),
  ].joined(separator: "\n")

  let kdate = hmac(Array(("AWS4" + input.secret).utf8), date)
  let kregion = hmac(kdate, input.region)
  let kservice = hmac(kregion, input.service)
  let ksigning = hmac(kservice, "aws4_request")
  let signature = hex(hmac(ksigning, stringtosign))

  var out = Ordered<String>()

  out["authorization"] =
    "AWS4-HMAC-SHA256 Credential=\(input.keyid)/\(scope)"
    + ", SignedHeaders=\(signedheaders)"
    + ", Signature=\(signature)"
  out["x-amz-date"] = input.datetime

  if let token = session { out["x-amz-security-token"] = token }

  return out
}

File: cpp/plugins/Sigv4.cpp
#include "Sigv4.hpp"

#include <algorithm>
#include <cstdio>

#include "Crypto.hpp"

namespace sekreto {

std::string sha256hex(const std::string& text) {
  return hex(sha256(tobytes(text)));
}

namespace {

Bytes hmactext(const Bytes& key, const std::string& text) {
  return hmacsha256(key, tobytes(text));
}

}  // namespace

std::string uriescape(const std::string& text) {
  std::string out;

  for (unsigned char byte : text) {
    bool unreserved = ('A' <= byte && 'Z' >= byte) || ('a' <= byte && 'z' >= byte) ||
                      ('0' <= byte && '9' >= byte) || '-' == byte || '_' == byte ||
                      '.' == byte || '~' == byte;

    if (unreserved) {
      out.push_back(static_cast<char>(byte));
    } else {
      char buf[8];
      std::snprintf(buf, sizeof(buf), "%%%02X", byte);
      out += buf;
    }
  }

  return out;
}

std::string uridecode(const std::string& text) {
  std::string out;
  size_t index = 0;

  while (index < text.size()) {
    bool taken = false;

    if ('%' == text[index] && index + 2 < text.size()) {
      std::string digits = text.substr(index + 1, 2);
      bool ok = true;
      unsigned int code = 0;

      for (char digit : digits) {
        unsigned int val = 0;
        if ('0' <= digit && '9' >= digit) {
          val = static_cast<unsigned int>(digit - '0');
        } else if ('a' <= digit && 'f' >= digit) {
          val = static_cast<unsigned int>(digit - 'a' + 10);
        } else if ('A' <= digit && 'F' >= digit) {
          val = static_cast<unsigned int>(digit - 'A' + 10);
        } else {
          ok = false;
          break;
        }
        code = (code << 4) | val;
      }

      if (ok) {
        out.push_back(static_cast<char>(code));
        index += 3;
        taken = true;
      }
    }

    if (!taken) {
      out.push_back(text[index]);
      index++;
    }
  }

  return out;
}

std::string canonicalquery(const std::string& query) {
  if (query.empty()) return "";

  std::vector<std::pair<std::string, std::string>> pairs;
  size_t start = 0;

  while (true) {
    size_t at = query.find('&', start);
    std::string pair =
        (std::string::npos == at) ? query.substr(start) : query.substr(start, at - start);

    // Split on the FIRST `=`; a pair with none has an empty value and is
    // still emitted with its `=`.
    size_t eq = pair.find('=');
    std::string key = (std::string::npos == eq) ? pair : pair.substr(0, eq);
    std::string value = (std::string::npos == eq) ? "" : pair.substr(eq + 1);

    pairs.emplace_back(uriescape(uridecode(key)), uriescape(uridecode(value)));

    if (std::string::npos == at) break;
    start = at + 1;
  }

  std::stable_sort(pairs.begin(), pairs.end(),
                   [](const std::pair<std::string, std::string>& left,
                      const std::pair<std::string, std::string>& right) {
                     if (left.first != right.first) return left.first < right.first;
                     return left.second < right.second;
                   });

  std::string out;
  for (size_t index = 0; index < pairs.size(); index++) {
    if (0 < index) out += "&";
    out += pairs[index].first + "=" + pairs[index].second;
  }

  return out;
}

namespace {

/// A URL, split by hand.
///
/// Not a platform URL type: what SigV4 signs must match what checkaddr saw
/// and what the client will dial, and a URL parser is free to normalise
/// any of the three differently. `host` follows the WHATWG rule the AWS
/// SDKs use - lowercased, userinfo stripped, the port present only when it
/// is not the scheme's default.
struct Urlparts {
  std::string scheme;
  std::string host;
  std::string path;
  std::string query;
};

Urlparts urlparts(const std::string& url) {
  Urlparts out;

  std::string rest = url;
  size_t mark = url.find("://");

  if (std::string::npos != mark) {
    out.scheme = asciilower(url.substr(0, mark));
    rest = url.substr(mark + 3);
  }

  std::string authority = rest;
  std::string tail;

  size_t stop = rest.find_first_of("/?#");
  if (std::string::npos != stop) {
    authority = rest.substr(0, stop);
    tail = rest.substr(stop);
  }

  // Userinfo is not part of the host, and never part of a signature.
  size_t at = authority.rfind('@');
  if (std::string::npos != at) authority = authority.substr(at + 1);

  std::string hostname = authority;
  std::string port;

  if (!authority.empty() && '[' == authority[0]) {
    size_t close = authority.find(']');
    if (std::string::npos != close) {
      hostname = authority.substr(0, close + 1);
      std::string after = authority.substr(close + 1);
      if (!after.empty() && ':' == after[0]) port = after.substr(1);
    }
  } else {
    size_t colon = authority.rfind(':');
    if (std::string::npos != colon) {
      hostname = authority.substr(0, colon);
      port = authority.substr(colon + 1);
    }
  }

  hostname = asciilower(hostname);

  // A default port is implicit: `Host: example.com:443` is not what an AWS
  // SDK signs, and a signature over the wrong host is refused.
  if (("https" == out.scheme && "443" == port) || ("http" == out.scheme && "80" == port)) {
    port = "";
  }

  out.host = port.empty() ? hostname : hostname + ":" + port;
  out.path = "/";

  if (!tail.empty()) {
    std::string pathpart = tail;

    size_t hash = pathpart.find('#');
    if (std::string::npos != hash) pathpart = pathpart.substr(0, hash);

    size_t question = pathpart.find('?');
    if (std::string::npos != question) {
      out.query = pathpart.substr(question + 1);
      pathpart = pathpart.substr(0, question);
    }

    if (!pathpart.empty()) out.path = pathpart;
  }

  return out;
}

/// Trim, and collapse every internal run of spaces and tabs to one space.
///
/// AWS folds sequential whitespace before signing, so a header value like
/// `a  b\tc` must sign as `a b c` or the service refuses the request.
std::string foldspace(const std::string& text) {
  std::string out;
  bool pending = false;
  bool started = false;

  for (char ch : text) {
    if (' ' == ch || '\t' == ch || '\n' == ch || '\r' == ch) {
      if (started) pending = true;
      continue;
    }
    if (pending) {
      out.push_back(' ');
      pending = false;
    }
    out.push_back(ch);
    started = true;
  }

  return out;
}

}  // namespace

Ordered sigv4(const Signing& input) {
  Urlparts parts = urlparts(input.url);

  std::string date = input.datetime.substr(0, std::min<size_t>(8, input.datetime.size()));

  // Every header that will be signed: the caller's extras, plus host and
  // x-amz-date (and the session token when present). The built-in three go
  // in AFTER the caller's, so they win over anything passed in.
  Ordered headers;

  for (const auto& entry : input.headers.pairs) {
    headers.set(asciilower(entry.first), foldspace(entry.second));
  }

  headers.set("host", parts.host);
  headers.set("x-amz-date", input.datetime);
  if (!input.session.empty()) headers.set("x-amz-security-token", input.session);

  std::vector<std::pair<std::string, std::string>> sorted = headers.pairs;
  std::stable_sort(sorted.begin(), sorted.end(),
                   [](const std::pair<std::string, std::string>& left,
                      const std::pair<std::string, std::string>& right) {
                     return left.first < right.first;
                   });

  std::string canonicalheaders;
  std::string signedheaders;

  for (size_t index = 0; index < sorted.size(); index++) {
    canonicalheaders += sorted[index].first + ":" + sorted[index].second + "\n";
    if (0 < index) signedheaders += ";";
    signedheaders += sorted[index].first;
  }

  // Six lines. `canonicalheaders` already ends with a newline, which is
  // what produces the blank line the algorithm calls for.
  std::string canonicalrequest = asciiupper(input.method) + "\n" + parts.path + "\n" +
                                 canonicalquery(parts.query) + "\n" + canonicalheaders + "\n" +
                                 signedheaders + "\n" + sha256hex(input.body);

  std::string scope = date + "/" + input.region + "/" + input.service + "/aws4_request";

  std::string stringtosign = "AWS4-HMAC-SHA256\n" + input.datetime + "\n" + scope + "\n" +
                             sha256hex(canonicalrequest);

  Bytes kdate = hmactext(tobytes("AWS4" + input.secret), date);
  Bytes kregion = hmactext(kdate, input.region);
  Bytes kservice = hmactext(kregion, input.service);
  Bytes ksigning = hmactext(kservice, "aws4_request");
  std::string signature = hex(hmactext(ksigning, stringtosign));

  Ordered out;

  out.set("authorization", "AWS4-HMAC-SHA256 Credential=" + input.keyid + "/" + scope +
                               ", SignedHeaders=" + signedheaders +
                               ", Signature=" + signature);
  out.set("x-amz-date", input.datetime);

  if (!input.session.empty()) out.set("x-amz-security-token", input.session);

  return out;
}

}  // namespace sekreto
